File: packages/health_connector_hc_android/test/unit_tests/src/mappers/health_record_mappers/mindfulness_session_record_mapper_test.dart
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/mindfulness_session_record_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';
import 'package:test/test.dart';

import '../../../utils/fake_data.dart';

void main() {
  group(
    'MindfulnessSessionRecordMapper',
    () {
      group(
        'MindfulnessSessionRecordToDto',
        () {
          test(
            'converts MindfulnessSessionRecord to DTO',
            () {
              final startTime = DateTime(2025, 1, 15, 10).toUtc();
              final endTime = DateTime(2025, 1, 15, 10, 20).toUtc();
              const title = 'Morning Meditation';
              const notes = 'Peaceful session';

              final record = MindfulnessSessionRecord(
                id: HealthRecordId(FakeData.fakeId),
                startTime: startTime,
                endTime: endTime,
                startZoneOffsetSeconds: 3_600,
                endZoneOffsetSeconds: 3_600,
                metadata: const Metadata(
                  dataOrigin: DataOrigin(FakeData.fakeDataOrigin),
                  recordingMethod: RecordingMethod.activelyRecorded,
                  clientRecordVersion: 1,
                  device: Device(type: DeviceType.phone),
                ),
                sessionType: MindfulnessSessionType.meditation,
                title: title,
                notes: notes,
              );

              final dto = record.toDto();

              expect(dto.id, FakeData.fakeId);
              expect(dto.startTime, startTime.millisecondsSinceEpoch);
              expect(dto.endTime, endTime.millisecondsSinceEpoch);
              expect(dto.metadata.dataOrigin, FakeData.fakeDataOrigin);
              expect(dto.sessionType, MindfulnessSessionTypeDto.meditation);
              expect(dto.title, title);
              expect(dto.notes, notes);
            },
          );
        },
      );

      group(
        'MindfulnessSessionRecordDtoToDomain',
        () {
          test(
            'converts DTO to MindfulnessSessionRecord',
            () {
              final startTime = DateTime(2025, 1, 15, 10).toUtc();
              final endTime = DateTime(2025, 1, 15, 10, 15).toUtc();
              const title = 'Breathing Exercise';
              const notes = 'Relaxing';

              final dto = MindfulnessSessionRecordDto(
                id: FakeData.fakeId,
                startTime: startTime.millisecondsSinceEpoch,
                endTime: endTime.millisecondsSinceEpoch,
                startZoneOffsetSeconds: 0,
                endZoneOffsetSeconds: 0,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.manualEntry,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                sessionType: MindfulnessSessionTypeDto.breathing,
                title: title,
                notes: notes,
              );

              final record = dto.toDomain();

              expect(record.id.value, FakeData.fakeId);
              expect(record.startTime, startTime);
              expect(record.endTime, endTime);
              expect(
                record.metadata.dataOrigin.packageName,
                FakeData.fakeDataOrigin,
              );
              expect(record.sessionType, MindfulnessSessionType.breathing);
              expect(record.title, title);
              expect(record.notes, notes);
            },
          );

          test(
            'converts DTO with null id to domain with none id',
            () {
              final startTime = DateTime(2025, 1, 15, 10).toUtc();
              final endTime = DateTime(2025, 1, 15, 10, 10).toUtc();

              final dto = MindfulnessSessionRecordDto(
                startTime: startTime.millisecondsSinceEpoch,
                endTime: endTime.millisecondsSinceEpoch,
                startZoneOffsetSeconds: 0,
                endZoneOffsetSeconds: 0,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.activelyRecorded,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                sessionType: MindfulnessSessionTypeDto.unknown,
              );

              final record = dto.toDomain();

              expect(record.id, HealthRecordId.none);
            },
          );
        },
      );
    },
  );
}

File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_record_mappers/MindfulnessSessionRecordMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_record_mappers

import androidx.health.connect.client.feature.ExperimentalMindfulnessSessionApi
import androidx.health.connect.client.records.MindfulnessSessionRecord
import androidx.health.connect.client.records.metadata.Metadata
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.DeviceTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.MetadataDto
import com.phamtunglam.health_connector_hc_android.pigeon.MindfulnessSessionRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.MindfulnessSessionTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.RecordingMethodDto
import io.kotest.matchers.shouldBe
import java.time.Instant
import java.time.ZoneOffset
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@OptIn(ExperimentalMindfulnessSessionApi::class)
@DisplayName("MindfulnessSessionRecordMapper")
class MindfulnessSessionRecordMapperTest {

    private companion object {
        const val TEST_ID = "mindfulness-id"
        const val TEST_START_TIME = 1609459200000L
        const val TEST_END_TIME = 1609462800000L
        const val TEST_TITLE = "Morning Meditation"
        const val TEST_NOTES = "Feeling calm"
        val TEST_ZONE_OFFSET: ZoneOffset = ZoneOffset.ofHours(1)
    }

    @Test
    @DisplayName("GIVEN MindfulnessSessionRecord → WHEN toDto → THEN converts correctly")
    fun whenRecordToDto_thenConvertsCorrectly() {
        // Given
        val record = MindfulnessSessionRecord(
            startTime = Instant.ofEpochMilli(TEST_START_TIME),
            endTime = Instant.ofEpochMilli(TEST_END_TIME),
            startZoneOffset = TEST_ZONE_OFFSET,
            endZoneOffset = TEST_ZONE_OFFSET,
            metadata = Metadata.manualEntry(),
            mindfulnessSessionType = MindfulnessSessionRecord.MINDFULNESS_SESSION_TYPE_MEDITATION,
            title = TEST_TITLE,
            notes = TEST_NOTES,
        )

        // When
        val result = record.toDto()

        // Then
        result.startTime shouldBe TEST_START_TIME
        result.endTime shouldBe TEST_END_TIME
        result.startZoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
        result.endZoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
        result.sessionType shouldBe MindfulnessSessionTypeDto.MEDITATION
        result.title shouldBe TEST_TITLE
        result.notes shouldBe TEST_NOTES
    }

    @Test
    @DisplayName(
        "GIVEN MindfulnessSessionRecordDto → WHEN toHealthConnect → THEN creates correctly",
    )
    fun whenDtoToHealthConnect_thenCreatesCorrectly() {
        // Given
        val dto = MindfulnessSessionRecordDto(
            id = TEST_ID,
            startTime = TEST_START_TIME,
            endTime = TEST_END_TIME,
            startZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            endZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            sessionType = MindfulnessSessionTypeDto.BREATHING,
            title = TEST_TITLE,
            notes = TEST_NOTES,
            metadata = MetadataDto(
                dataOrigin = "com.example.app",
                recordingMethod = RecordingMethodDto.MANUAL_ENTRY,
                deviceType = DeviceTypeDto.PHONE,
            ),
        )

        // When
        val result = dto.toHealthConnect()

        // Then
        result.startTime shouldBe Instant.ofEpochMilli(TEST_START_TIME)
        result.endTime shouldBe Instant.ofEpochMilli(TEST_END_TIME)
        result.startZoneOffset shouldBe TEST_ZONE_OFFSET
        result.endZoneOffset shouldBe TEST_ZONE_OFFSET
        result.mindfulnessSessionType shouldBe
            MindfulnessSessionRecord.MINDFULNESS_SESSION_TYPE_BREATHING
        result.title shouldBe TEST_TITLE
        result.notes shouldBe TEST_NOTES
        result.metadata.id shouldBe TEST_ID
    }
}
